packages/office-ui-fabric-react/src/components/SpinButton/SpinButton.tsx

```diff
@@ -0,0 +1,408 @@
+   * when spinning in response to a mouseDown
+   * @param stepFunction - function to use to step by
+   */
+  @autobind
+  private _updateValue(shouldSpin: boolean, stepFunction: (string) => string | void) {
+    const newValue = stepFunction(this.state.value);
+    if (newValue) {
+      this._lastValidValue = newValue;
+      this.setState({ value: newValue });
+    }
+
+    if (this._spinningByMouse !== shouldSpin) {
+      this._spinningByMouse = shouldSpin;
+    }
+
+    if (shouldSpin) {
+      this._currentStepFunctionHandle = this._async.setTimeout(() => { this._updateValue(shouldSpin, stepFunction); }, this._stepDelay);
+    }
+  }
+
+  /**
+   * Stop spinning (clear any currently pending update and set spinning to false)
+   */
+  @autobind
+  private _stop() {
+    if (this._currentStepFunctionHandle !== null) {
+      this._async.clearTimeout(this._currentStepFunctionHandle);
+      this._currentStepFunctionHandle = 0;
+    }
+
+    if (this._spinningByMouse || this.state.keyboardSpinDirection !== KeyboardSpinDirection.notSpinning) {
+      this._spinningByMouse = false;
+      this.setState({ keyboardSpinDirection: KeyboardSpinDirection.notSpinning });
+    }
+  }
+
+  /**
+   * Handle keydown on the text field. We need to update
+   * the value when up or down arrow are depressed
+   * @param event - the keyboardEvent that was fired
+   */
+  @autobind
+  private _handleKeyDown(event: React.KeyboardEvent<HTMLElement>) {
+    if (this.props.disabled) {
+      this._stop();
+
+      // eat the up and down arrow keys to keep the page from scrolling
+      if (event.which === KeyCodes.up || event.which === KeyCodes.down) {
+        event.preventDefault();
+        event.stopPropagation();
+      }
+
+      return;
+    }
+
+    let spinDirection = KeyboardSpinDirection.notSpinning;
+
+    if (event.which === KeyCodes.up) {
+
+      spinDirection = KeyboardSpinDirection.up;
+      this._updateValue(false /* shouldSpin */, this._onIncrement);
+    } else if (event.which === KeyCodes.down) {
+
+      spinDirection = KeyboardSpinDirection.down;
+      this._updateValue(false /* shouldSpin */, this._onDecrement);
+    } else if (event.which === KeyCodes.enter) {
+      event.currentTarget.blur();
+      this.focus();
+    } else if (event.which === KeyCodes.escape) {
+      if (this.state.value !== this._lastValidValue) {
+        this.setState({ value: this._lastValidValue });
+      }
+    }
+
+    // style the increment/decrement button to look active
+    // when the corresponding up/down arrow keys trigger a step
+    if (this.state.keyboardSpinDirection !== spinDirection) {
+      this.setState({ keyboardSpinDirection: spinDirection });
+    }
+  }
+
+  /**
+   * Make sure that we have stopped spinning on keyUp
+   * if the up or down arrow fired this event
+   * @param event stop spinning if we
+   */
+  @autobind
+  private _handleKeyUp(event: React.KeyboardEvent<HTMLElement>) {
+
+    if (this.props.disabled) {
```

Comment:
Sorry about the missing comments, sometimes GitHub collapses comments for no apparent reason, making it hard to see what has been addressed or not.